### PR TITLE
Added Elementwise division

### PR DESCRIPTION
**Pull request by Khalifa1997**
Added elementwise division with [Eigen's division operator](https://eigen.tuxfamily.org/dox/classEigen_1_1VectorwiseOp.html#aa154a8407885a9ad51504180f6d7a163) one thing to note however, the case whereby there is division by zero isn't yet handled

### CODE REVIEW

**Comment by gf712** [comment]
I guess you get NaN values in zero division? That is the behaviour in C++ I think.
I am not sure I understand why you have the transpose versions?

**Comment by Khalifa1997** [comment]
> I guess you get NaN values in zero division? That is the behaviour in C++ I think.

You are right, It does evaluate to NaN well or -NaN.

> I am not sure I understand why you have the transpose versions?

I wanted to stay consistent with `element_prod` as it also takes transpose as an option, and it's a good feature to add imo

**Comment by gf712** [comment]
> > I guess you get NaN values in zero division? That is the behaviour in C++ I think.
> 
> You are right, It does evaluate to NaN well or -NaN.

Actually it is undefined behaviour, so Eigen will populate the zero division results with an internal representation of `inf`, see https://godbolt.org/z/-WAuL-. @karlnapf should linalg throw? I don't know what happens if for example you invert a singular matrix? In that case you would also have zero division no?

> > I am not sure I understand why you have the transpose versions?
> 
> I wanted to stay consistent with `element_prod` as it also takes transpose as an option, and it's a good feature to add imo

Only if absolutely necessary.. You are adding a lot of code, that takes a long time to compile.. @karlnapf ?

**Comment by Khalifa1997** [comment]
> > > I guess you get NaN values in zero division? That is the behaviour in C++ I think.
> > 
> > 
> > You are right, It does evaluate to NaN well or -NaN.
> 
> Actually it is undefined behaviour, so Eigen will populate the zero division results with an internal representation of `inf`, see https://godbolt.org/z/-WAuL-. @karlnapf should linalg throw? I don't know what happens if for example you invert a singular matrix? In that case you would also have zero division no?
> 
Hmm interesting, I think this can be done by an if condition before division like this, `bool has_zero = (B.array() == 0.0).any();` then I would use the error function incase this line is true

**Comment by gf712** [comment]
> > > > I guess you get NaN values in zero division? That is the behaviour in C++ I think.
> > > 
> > > 
> > > You are right, It does evaluate to NaN well or -NaN.
> > 
> > 
> > Actually it is undefined behaviour, so Eigen will populate the zero division results with an internal representation of `inf`, see https://godbolt.org/z/-WAuL-. @karlnapf should linalg throw? I don't know what happens if for example you invert a singular matrix? In that case you would also have zero division no?
> 
> Hmm interesting, I think this can be done by an if condition before division like this, `bool has_zero = (B.array() == 0.0).any();` then I would use the error function incase this line is true

Yes, but then everyone has to pay the price of that check.. What if I am dividing by something I know can't be zero? For example divide by a vector of distances?

**Comment by Khalifa1997** [comment]
> > > > > I guess you get NaN values in zero division? That is the behaviour in C++ I think.
> > > > 
> > > > 
> > > > You are right, It does evaluate to NaN well or -NaN.
> > > 
> > > 
> > > Actually it is undefined behaviour, so Eigen will populate the zero division results with an internal representation of `inf`, see https://godbolt.org/z/-WAuL-. @karlnapf should linalg throw? I don't know what happens if for example you invert a singular matrix? In that case you would also have zero division no?
> > 
> > 
> > Hmm interesting, I think this can be done by an if condition before division like this, `bool has_zero = (B.array() == 0.0).any();` then I would use the error function incase this line is true
> 
> Yes, but then everyone has to pay the price of that check.. What if I am dividing by something I know can't be zero? For example divide by a vector of distances?

Yea, I understand.. we will see what @karlnapf thinks

File: src/shogun/mathematics/linalg/LinalgBackendViennaCL.h [hunk 1]
```diff
@@ -399,7 +409,45 @@ namespace shogun
 				        a_gpu->data_matrix(a.num_rows, a.num_cols),
 				        b_gpu->data_matrix(b.num_rows, b.num_cols));
 		}
+		/** ViennaCL matrix in-place elementwise division method */
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendViennaCL.h]: no need to populate this, as it doesnt work anyways

File: src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp [hunk 2]
```diff
@@ -446,6 +477,63 @@ void LinalgBackendEigen::element_prod_impl(
 	result_eig = a_eig.array() * b_eig.array();
 }
 
+/* Helper method to compute elementwise division with Eigen */
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp:489]: @gf712 just another illustration for the fact that we need a data structure for linear algebra. That can for example have a ".T" flag inside :)

**Comment by karlnapf** [comment]
IMO we don't need the transpose stuff. But I agree it is good to stay consistent with what we have. Not sure what to do. In terms of dividing by zero I am clearly in favor of NaN

**Comment by karlnapf** [comment]
lgtm in general, given that the current linalg defines things like this.

**Comment by gf712** [comment]
> IMO we don't need the transpose stuff. But I agree it is good to stay consistent with what we have. Not sure what to do. In terms of dividing by zero I am clearly in favor of NaN

@karlnapf what do you mean with the second part? That we should not throw with zero division and assume the caller knows the risk?

**Comment by gf712** [comment]
> lgtm in general, given that the current linalg defines things like this.

My issue is that with my 16GB RAM laptop I struggle already with compiling shogun in debug mode. All the transpose combinations will make shogun uncompilable soon, for some ppl at least..

File: src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp (hunk 2 above)
Comment (gf712) [reply src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp:489]: yup, like eigen.. Something being worked on in the graph stuff!

**Comment by karlnapf** [comment]
Actually just checked and python throws an error, but numpy array operations return inf. I think following that is good

**Comment by karlnapf** [comment]
let's not have the transpose then. It is not needed, and actually doesnt even make that much sense for element wise operations

**Comment by Khalifa1997** [comment]
> let's not have the transpose then. It is not needed, and actually doesnt even make that much sense for element wise operations

So remove transpose and keep the NaN?

File: src/shogun/mathematics/linalg/LinalgBackendViennaCL.h (hunk 1 above)
Comment (karlnapf) [reply src/shogun/mathematics/linalg/LinalgBackendViennaCL.h]: i.e. remove

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 3]
```diff
@@ -1156,6 +1156,165 @@ namespace shogun
...
+			    "Dimension mismatch! A({} x {}) vs B({} x {})", A.m_row_size,
+			    A.m_col_size, B.m_row_size, B.m_col_size);
+
+			require(
+			    A.m_row_size == result.num_rows && A.m_col_size == result.num_cols,
+			    "Dimension mismatch! A({} x {}) vs result({} x {})",
+			    A.m_row_size, A.m_col_size, result.num_rows, result.num_cols);
+
+			require(
+			    !result.on_gpu(),
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: @gf712 as we disabled this iirc. What about a quick refactor where all `.on_gpu` calls are removed?

File: src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp (hunk 2 above)
Comment (karlnapf) [reply src/shogun/mathematics/linalg/backend/eigen/BasicOps.cpp:489]: :rocket:

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 4]
```diff
@@ -1002,6 +1002,117 @@ TYPED_TEST(
...
+	{
+		a[i] = i;
+		b[i] = i + 1;
+		c[i] = i;
+	}
+
+	element_div(a, b, a);
+	for (index_t i = 0; i < len; ++i)
+		EXPECT_NEAR(c[i] / b[i], a[i], get_epsilon<TypeParam>());
+}
```
Comment (karlnapf) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1121]: pls also test the corner cases where matrix dimensions do not match, or where inf is returned (can be on very small minimal element)

File: src/shogun/mathematics/linalg/LinalgNamespace.h (hunk 3 above)
Comment (gf712) [reply src/shogun/mathematics/linalg/LinalgNamespace.h]: we still have it. What we removed were the calls from SGVector/Matrix if ViennaCL is not used to compile the library

File: src/shogun/mathematics/linalg/LinalgBackendBase.h [hunk 5]
```diff
@@ -404,6 +404,55 @@ namespace shogun
...
+
+/**
+ * Wrapper method of in-place matrix block elementwise division.
+ *
+ * @see linalg::element_div
+ */
+#define BACKEND_GENERIC_IN_PLACE_BLOCK_ELEMENT_DIV(Type, Container)           \
+	virtual void element_div(                                                 \
+	    const linalg::Block<Container<Type>>& a,                               \
+	    const linalg::Block<Container<Type>>& b, Container<Type>& result) const                              \
```
Comment (gf712) [inline src/shogun/mathematics/linalg/LinalgBackendBase.h]: Do we need/use these?

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc (hunk 4 above)
Comment (Khalifa1997) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1121]: I am using something like this
`EXPECT_THROW(element_div(SGMatrix<TypeParam> (m+1, m), d), ShogunException);`
or would you prefer that or should I use smth else

File: src/shogun/mathematics/linalg/LinalgBackendBase.h (hunk 5 above)
Comment (Khalifa1997) [reply src/shogun/mathematics/linalg/LinalgBackendBase.h]: I wanted to remain consistent with `element_prod`.. what do you think @karlnapf

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 6]
```diff
@@ -1002,6 +1002,159 @@ TYPED_TEST(
...
+{
+	const auto m = 2;
+	SGMatrix<TypeParam> A(m, m);
+	SGMatrix<TypeParam> B(m, m);
+
+	for (auto i : range(m * m))
+	{
+		A[i] = i;
+		B[i] = i + 1;
+	}
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: what about `std::iota`? cleans this up a bit

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 7]
```diff
@@ -1002,6 +1002,159 @@ TYPED_TEST(
...
+	const auto m = 2;
+	SGMatrix<TypeParam> A(m, m);
+	SGMatrix<TypeParam> B(m, m);
+	SGMatrix<TypeParam> result(m, m);
+
+	for (auto i : range(m * m))
+	{
+		A[i] = i;
+		B[i] = i + 1;
+	}
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: `std::iota`

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 8]
```diff
@@ -1002,6 +1002,159 @@ TYPED_TEST(
...
+	for (auto i : range(nrows))
+		for (auto j : range(ncols))
+		{
+			A(i, j) = i * 10 + j + 1;
+			B(j, i) = i + j + 1;
+		}
+
+	const auto m = 2;
+	auto A_block = linalg::block(A, 0, 0, m, m);
+	auto B_block = linalg::block(B, 0, 0, m, m);
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: are you going to need these block operations? I think these should be added when needed, otherwise it's just binary bloat

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 9]
```diff
@@ -1002,6 +1002,159 @@ TYPED_TEST(
...
+	const index_t len = 4;
+	SGVector<TypeParam> a(len);
+	SGVector<TypeParam> b(len);
+	SGVector<TypeParam> c(len);
+
+	for (index_t i = 0; i < len; ++i)
+	{
+		a[i] = i;
+		b[i] = i + 1;
+	}
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: `range_fill()`

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 10]
```diff
@@ -1002,6 +1002,159 @@ TYPED_TEST(
...
+	SGVector<TypeParam> a(len);
+	SGVector<TypeParam> b(len);
+	SGVector<TypeParam> c(len);
+
+	for (index_t i = 0; i < len; ++i)
+	{
+		a[i] = i;
+		b[i] = i + 1;
+		c[i] = i;
+	}
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: `range_fill`

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc (hunk 8 above)
Comment (Khalifa1997) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: so do you recommend I remove the whole thing? aka the block `element_div`?

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 11]
```diff
@@ -1002,6 +1002,159 @@ TYPED_TEST(
...
+
+	element_div(a, b, a);
+	for (index_t i = 0; i < len; ++i)
+		EXPECT_NEAR(c[i] / b[i], a[i], get_epsilon<TypeParam>());
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: btw, there is a `zip_iterator`, so you can write this python style
```cpp
for (const auto& [result, el1, el2]: zip_iterator(a,b,c))
    EXPECT_NEAR(el2 / el1, result, get_epsilon<TypeParam>());
```
Same in all the other loops where you check the result. Another thing that you should try to use in your code, because it cleans it up a lot

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc (hunk 8 above)
Comment (gf712) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: yes, unless you need it ofc! :)

Comment (gf712) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: but can wait for @karlnapf input or maybe @vigsterkr has thoughts on this?

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 12]
```diff
@@ -1002,6 +1003,145 @@ TYPED_TEST(
...
+TYPED_TEST(
+    LinalgBackendEigenAllTypesTest, SGMatrix_elementwise_division_in_place)
+{
+	const auto m = 2;
+	SGMatrix<TypeParam> A(m, m);
+	SGMatrix<TypeParam> B(m, m);
+	SGMatrix<TypeParam> result(m, m);
+
+	std::iota(A.matrix, A.matrix + m*2, 0);
+	std::iota(B.matrix, B.matrix + m*2, 1);
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: why `m*2`? surely `m*m`.. Also, you should get the matrix length with `SGMatrix::size()`

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 13]
```diff
@@ -1002,6 +1003,145 @@ TYPED_TEST(
...
 		EXPECT_NEAR(c[i] * b[i], a[i], get_epsilon<TypeParam>());
 }
+TYPED_TEST(LinalgBackendEigenAllTypesTest, SGMatrix_elementwise_division)
+{
+	const auto m = 2;
+	SGMatrix<TypeParam> A(m, m);
+	SGMatrix<TypeParam> B(m, m);
+
+	std::iota(A.matrix, A.matrix + m*2, 0);
+	std::iota(B.matrix, B.matrix + m*2, 1);
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: use `SGMatrix::size()`

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc (hunk 12 above)
Comment (gf712) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: I guess you put `m*2` because `m*2==m*m` in this case... But this is can lead to bugs in refactors, and is also unclear

**Comment by Khalifa1997** [comment]
This is quite strange, even though the bits of code I had added which should theoretically handle division by zero, it just breaks the whole tests even creating test errors in stuff like elementwise division of matrices of different sizes -even though those particular tests have passed before adding the division by zero test- is there anything I am missing here?.. @gf712 @karlnapf

**Comment by gf712** [comment]
> This is quite strange, even though the bits of code I had added which should theoretically handle division by zero, it just breaks the whole tests even creating test errors in stuff like elementwise division of matrices of different sizes -even though those particular tests have passed before adding the division by zero test- is there anything I am missing here?.. @gf712 @karlnapf

OK, first, the other tests are still passing. 
Second, the issue is that zero division is only defined for floating point types (that's why those test pass if you look at the CI output). Integer types are undefined and therefore throw.
You should add a `if constexpr` to check that the test type is floating point type.

**Comment by Khalifa1997** [comment]
@gf712 is this good to merge? I believe the tests only failed cause it couldn't pull the docker image correctly

**Comment by gf712** [comment]
> @gf712 is this good to merge? I believe the tests only failed cause it couldn't pull the docker image correctly

Just waiting for @karlnapf's input about the block operations

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 14]
```diff
@@ -1002,6 +1004,158 @@ TYPED_TEST(
...
+	EXPECT_THROW(element_div(C, A, B), ShogunException);
+	EXPECT_THROW(element_div(D, A, B), ShogunException);
+	EXPECT_THROW(element_div(E, A, B), ShogunException);
+
+	if constexpr (std::is_floating_point<TypeParam>::value){
+	B.zero();
+	element_div(A, B, result);
+	typename SGMatrix<TypeParam>::EigenMatrixXtMap result_eig = result;
+    	EXPECT_TRUE(result_eig.array().isInf().any() && result_eig.array().isNaN().any());
+	}
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1071]: Could you indent this please

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 15]
```diff
@@ -1002,6 +1004,158 @@ TYPED_TEST(
...
+
+	for (const auto& [result, el1, el2]: zip_iterator(c,b,a))
+    		EXPECT_NEAR(el2 / el1, result, get_epsilon<TypeParam>());
+
+	if constexpr (std::is_floating_point<TypeParam>::value){
+	b.zero();
+	c = element_div(a, b);
+	typename SGVector<TypeParam>::EigenRowVectorXtMap result_eig = c;
+    	EXPECT_TRUE(result_eig.array().isInf().any() && result_eig.array().isNaN().any());
+	}
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1094]: indent

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 16]
```diff
@@ -1156,6 +1156,165 @@ namespace shogun
...
+		 * @return The result of the operation
+		 */
+		template <typename T>
+		SGVector<T> element_div(const SGVector<T>& a, const SGVector<T>& b)
+		{
+			require(
+			    a.vlen == b.vlen, "Dimension mismatch! A({}) vs B({})",
+			    a.vlen, b.vlen);
+
+			SGVector<T> result = a.clone();
```
Comment (gf712) [inline src/shogun/mathematics/linalg/LinalgNamespace.h:1251]: @karlnapf @vigsterkr  wouldn't it be cheaper in these situations to just allocate an empty vector/matrix? This is an allocation + memcpy, and the copied results will be overwritten

Comment (Khalifa1997) [reply src/shogun/mathematics/linalg/LinalgNamespace.h:1251]: I saw the `.clone` in my linalg functions if convenient I could issue a PR to clean them all

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc (hunk 8 above)
Comment (karlnapf) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: do it

Comment (karlnapf) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: we just added those for some specific cases we had

File: src/shogun/mathematics/linalg/LinalgNamespace.h (hunk 16 above)
Comment (karlnapf) [reply src/shogun/mathematics/linalg/LinalgNamespace.h:1251]: yes! This is actually something we should fix shogun wide!

Comment (karlnapf) [reply src/shogun/mathematics/linalg/LinalgNamespace.h:1251]: #5014

Comment (karlnapf) [reply src/shogun/mathematics/linalg/LinalgNamespace.h:1251]: #5014

**Review by karlnapf — COMMENTED** [review]
LGTM

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc (hunk 15 above)
Comment (gf712) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1094]: just this missing

**Review by gf712 — CHANGES_REQUESTED** [review]
Almost ready to merge!

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc (hunk 15 above)
Comment (Khalifa1997) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1094]: Do I remove block operation as @karlnapf suggested?

Comment (karlnapf) [reply tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1094]: yes let's do that

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 17]
```diff
@@ -1156,6 +1156,125 @@ namespace shogun
...
+		SGMatrix<T> element_div(
+		    const Block<SGMatrix<T>>& A, const Block<SGMatrix<T>>& B)
+		{
+
+			SGMatrix<T> result(A.m_row_size, A.m_col_size);
+
+			element_div(A, B, result);
+
+			return result;
+		}
```
Comment (gf712) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: I thought we decided to remove this?

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 18]
```diff
@@ -1002,6 +1004,121 @@ TYPED_TEST(
...
+	EXPECT_THROW(element_div(C, A, B), ShogunException);
+	EXPECT_THROW(element_div(D, A, B), ShogunException);
+	EXPECT_THROW(element_div(E, A, B), ShogunException);
+
+	if constexpr (std::is_floating_point<TypeParam>::value){
+	B.zero();
+	element_div(A, B, result);
+	typename SGMatrix<TypeParam>::EigenMatrixXtMap result_eig = result;
+    		EXPECT_TRUE(result_eig.array().isInf().any() && result_eig.array().isNaN().any());
+	}
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1071]: Could you please indent the whole block? This is just like any other `if` branch.

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 19]
```diff
@@ -1002,6 +1004,121 @@ TYPED_TEST(
...
+	c = element_div(a, b);
+
+	for (const auto& [result, el1, el2]: zip_iterator(c,b,a))
+    		EXPECT_NEAR(el2 / el1, result, get_epsilon<TypeParam>());
+
+	if constexpr (std::is_floating_point<TypeParam>::value){
+	b.zero();
+	c = element_div(a, b);
+	typename SGVector<TypeParam>::EigenRowVectorXtMap result_eig = c;
+    		EXPECT_TRUE(result_eig.array().isInf().any() && result_eig.array().isNaN().any());
```
Comment (gf712) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc:1093]: same here

File: src/shogun/mathematics/linalg/LinalgNamespace.h (hunk 17 above)
Comment (Khalifa1997) [reply src/shogun/mathematics/linalg/LinalgNamespace.h]: mb must of not seen it, fixed it
